Segresnet improvements

myron opened the pull request:
This adds improvements to SegResNet semantic segmentation to include deep supervision feature (among other things). 

I've decided to add a new network named SegResnetDS instead of replacing the existing SegResNet, because it would have broken all previously pretrained model checkpoints.

Please take a look, and if it looks good, I can add unit tests.   All the helper functions, and blocks are inside of the same file (which I prefer, since it's easier to read) instead of spreading it among other files (e.g segresnet_ds_block or segresnet_utils).  


### Description

A few sentences describing the changes proposed in this pull request.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

wyli commented on monai/networks/nets/segresnet_ds.py:
```diff
@@ -0,0 +1,476 @@
+__all__ = ["SegResNetDS"]
+
+
+def get_norm_layer(name: Union[Tuple, str], spatial_dims: Optional[int] = 1, channels: Optional[int] = 1):
```
Could you please reuse the existing api? or help improve the existing ones

myron replied:
I have tried to update existing API, e.g. for get_norm_layer() here is my PR https://github.com/Project-MONAI/MONAI/pull/5220

But it seems to have no support, since it'll trigger changes for all legacy models.  looks liek it's easier to add self-contained methods here, to prevent issues with legacy dependencies.

wyli replied:
> looks liek it's easier to add self-contained methods here, to prevent issues with legacy dependencies.

it's easier for this PR but adds maintenance burdens if we all follow this pattern of adding different variants of the same thing...

myron replied:
Sure, but what's the solution. Would you like to approve https://github.com/Project-MONAI/MONAI/pull/5220, then I can reuse existing functions, but it'll break previous models.

myron replied:
Any other comments/suggestions? If I do change get_norm_layer() and get_act_layer() for the main API it will break the functionality for legacy networks (not just segresnet).   https://github.com/Project-MONAI/MONAI/pull/5220

It seems it will simplify your work not to retest everything due to this small change. I prefer the model to be self-contained.

myron replied:
Following your suggestions, I've changed the code to use existing  get_norm_layer, get_act_layer, and set the defaults for my code inside of the network itself.

myron commented:
@wyli  I've updated the code to reuse shared get_norm_layer, get_act_layer, following your  request.
I've also added unit tests.  and internal blocks and components  are inside of the same file (instead of creating extra files), which I very much prefer.   Please check and merge.

wyli commented:
thanks, this looks to me, please address these minor issues detected: https://deepsource.io/gh/Project-MONAI/MONAI/run/2c41e145-4caa-4a5d-b7b9-78fa88b415ee/python/

for the torchscript issues (https://github.com/Project-MONAI/MONAI/actions/runs/3237807871/), the test could be skipped with a decorator:
https://github.com/Project-MONAI/MONAI/blob/16a0614b7ec724fe8e7fd91fb8e597ad1aa653c8/tests/test_dints_network.py#L153

myron commented:
> thanks, this looks to me, please address these minor issues detected: https://deepsource.io/gh/Project-MONAI/MONAI/run/2c41e145-4caa-4a5d-b7b9-78fa88b415ee/python/
> 
> for the torchscript issues (https://github.com/Project-MONAI/MONAI/actions/runs/3237807871/), the test could be skipped with a decorator:
> 
> https://github.com/Project-MONAI/MONAI/blob/16a0614b7ec724fe8e7fd91fb8e597ad1aa653c8/tests/test_dints_network.py#L153


Thanks, I've addressed all those issues, and pushed. plz check

wyli commented:
/build

nhhung1810 commented:
Hi @myron , I'm working on a PR about writing tutorials on this module `SegResNetDS`, can I ask some details, about
- Why this was an improvement compare to SegResnet and SegResNetVAE, because I can not find any papers that discuss about the SegResNet with deep supervision and non-isotropic kernel
- I wonder if this deep-supervision is the same as deep-supervision as UNet++, as I found the implementation quite similar
- It would be very helpful if you could point out some resources to non-isotropic/isotropic kernel, too, so that I can write the tutorials better 

Thank for your help :D
